internal/pkg/manifest/svc.go

```diff
@@ -442,3 +442,62 @@ func (s *stringOrFromCFN) UnmarshalYAML(value *yaml.Node) error {
 	}
 	return nil
 }
+
+func (cfg ImageWithPortAndHealthcheck) exposedPorts(workloadName string) (exposedPort []ExposedPort) {
+	// Read `image.port`
+	if cfg.Port != nil {
+		exposedPort = append(exposedPort, ExposedPort{
+			Port:          aws.Uint16Value(cfg.Port),
+			Protocol:      "tcp",
+			ContainerName: workloadName,
+		})
+	}
+	return exposedPort
+}
+
+func (cfg ImageWithHealthcheckAndOptionalPort) exposedPorts(workloadName string) (exposedPort []ExposedPort) {
+	// Read `image.port`
+	if cfg.Port != nil {
+		exposedPort = append(exposedPort, ExposedPort{
+			Port:          aws.Uint16Value(cfg.Port),
+			Protocol:      "tcp",
+			ContainerName: workloadName,
+		})
+	}
+	return exposedPort
+}
+
+func (rr RoutingRuleConfiguration) exposedPorts(workloadName string) (exposedPort []ExposedPort) {
+	// Read `http.target_port`
+	// This block of code takes care of following use cases:
+	// 1. if target_port is given but target_container is nil then set ContainerName as primary container.
+	// 2. if target_port is given with the target_container as primary or sidecar container then set the value to ContainerName.
+	containerName := workloadName
+	if rr.TargetPort != nil {
+		if rr.TargetContainer != nil {
+			containerName = aws.StringValue(rr.TargetContainer)
+		}
+		exposedPort = append(exposedPort, ExposedPort{
+			Port:          aws.Uint16Value(rr.TargetPort),
+			Protocol:      "tcp",
+			ContainerName: containerName,
+		})
+	}
+	return exposedPort
+}
+
+func (sidecar SidecarConfig) exposedPorts(sidecarName string) (exposedPort []ExposedPort) {
+	if sidecar.Port != nil {
+		sidecarPort, protocol, _ := ParsePortMapping(sidecar.Port)
```

Comment:
is it not possible that `sidecarPort` is `nil` on error, should we skip and return nil if it's malformed?

Reply:
So here and below [here](https://github.com/aws/copilot-cli/pull/4268#discussion_r1048995639) this same implementation is validated inside validation and returns error if port is malformed. So we do not need to worry here as validation has already taken care of those scenarios. Here is [validation code](https://github.com/aws/copilot-cli/pull/4268/files#diff-5fc13795122fe030799a09ef956a8b66cec3d40bf1a81266323edbf3626b5685R1752-R1757) for your reference.

Reply:
Gotcha!

I'm just always paranoid that something will go wrong if the validation step gets skipped :mildpanic-intensifies:
(this has nothing to do with your PR): We've been creating temporal coupling between these methods where it's assume that `Validate` is invoked before anything.

Perhaps, this is fine - would you mind adding a comment to the `ExposedPorts` methods saying that `Validate` should be invoked before calling this method?